### PR TITLE
ZCS-8756: ABQ - Create Admin API for add, modify, remove ABQ devices

### PR DESCRIPTION
**Pull request by nileshpawar1**
Issue:  Create Admin API for add, modify, remove ABQ devices

Fix:
Created SOAP API to add, modify and remove ABQ devices.

Testing done: 
Tested with SOAP UI tool with add, modify and remove operations

ABQDeviceRequest:

```
<soapenv:Envelope xmlns:soapenv="http://schemas.xmlsoap.org/soap/envelope/" xmlns:urn="urn:zimbra" xmlns:urn1="urn:zimbraAdmin">
   <soapenv:Header>
      <urn:context >
<urn:authToken>0_db9212d46ebd0b6a19c64da33c039b92c489d0d7_69643d33363a64316335393438382d346133652d343566612d383232652d6639303334316166353239353b6578703d31333a313538333831393933303537333b61646d696e3d313a313b747970653d363a7a696d6272613b753d313a613b7469643d31303a323039383539303235383b76657273696f6e3d31343a382e382e31355f47415f333032393b637372663d313a313b</urn:authToken>
      </urn:context>
   </soapenv:Header>
   <soapenv:Body>
      <urn1:ABQDeviceRequest op="add" deviceId="abc" accountId="abc-1" status="allowed" />
   </soapenv:Body>
</soapenv:Envelope>
```

### CODE REVIEW

File: common/src/java/com/zimbra/common/soap/AdminConstants.java [hunk 1]
```diff
@@ -532,6 +535,14 @@
     public static final String E_GET_ALL_SKINS_REQUEST = "GetAllSkinsRequest";
     public static final String E_GET_ALL_SKINS_RESPONSE = "GetAllSkinsResponse";
 
+    //ABQ
+    public static final String E_ABQ_DEVICE_REQUEST = "ABQDeviceRequest";
+    public static final String E_ABQ_DEVICE_RESPONSE = "ABQDeviceResponse";
+    public static final String E_DEVICE_ID = "deviceId";
```
Comment (dasiyogesh) [inline common/src/java/com/zimbra/common/soap/AdminConstants.java]: Can we rename the constant as E_ABQ_DEVICE_ID

File: soap/src/java/com/zimbra/soap/admin/message/ABQDeviceRequest.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    @XmlEnum
+    public enum ABQDeviceOperation {
+        // case must match
+        add, modify, remove;
+
+        public static ABQDeviceOperation fromString(String s) throws ServiceException {
+            try {
+                return ABQDeviceOperation.valueOf(s);
+            } catch (IllegalArgumentException e) {
+                throw ServiceException.INVALID_REQUEST("unknown key: "+s, e);
```
Comment (dasiyogesh) [inline soap/src/java/com/zimbra/soap/admin/message/ABQDeviceRequest.java]: add space in concatenation or better use String.format

File: store/src/java/com/zimbra/cs/db/DbMailbox.java [hunk 3]
```diff
@@ -1297,4 +1303,106 @@ private static void readMailboxRawData(List<Mailbox.MailboxData> results, Result
...
+                rs =  stmt.executeQuery();
+                if (rs.next()) {
+                    throw ServiceException.FAILURE(String.format("Db entry already exist for device_id = %s with account_id = %s", deviceId, accountId), null);
+                }
+                stmt.close();
+
+                stmt = conn.prepareStatement("INSERT INTO abq_devices(device_id, account_id, status, created_by, created_time) VALUES (?, ?, ?, ?, NOW())");
+                stmt.setString(1, deviceId);
+                stmt.setString(2, accountId);
+                stmt.setString(3, status);
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/db/DbMailbox.java]: can we add check for valid values of status before updating

File: store/src/java/com/zimbra/cs/db/DbMailbox.java [hunk 4]
```diff
@@ -1297,4 +1303,106 @@ private static void readMailboxRawData(List<Mailbox.MailboxData> results, Result
...
+            if (status == null || ABQ_DEVICE_STATUS.getAllowedStatus().contains(status)) {
+                conn = DbPool.getConnection();
+                if (accountId == null) {
+                    stmt = conn.prepareStatement("UPDATE abq_devices SET status = ?, modified_by = ?, modified_time = NOW() WHERE device_id = ? AND account_id IS NULL");
+                    stmt.setString(1, status);
+                    stmt.setString(2, modifiedBy);
+                    stmt.setString(3, deviceId);
+                } else {
+                    stmt = conn.prepareStatement("UPDATE abq_devices SET status = ?, modified_by = ?, modified_time = NOW() WHERE device_id = ? AND account_id = ?");
+                    stmt.setString(1, status);
```
Comment (dasiyogesh) [inline store/src/java/com/zimbra/cs/db/DbMailbox.java]: same here, add check for valid values for status

File: common/src/java/com/zimbra/common/soap/AdminConstants.java (hunk 1 above)
Comment (nileshpawar1) [reply common/src/java/com/zimbra/common/soap/AdminConstants.java]: done

File: soap/src/java/com/zimbra/soap/admin/message/ABQDeviceRequest.java (hunk 2 above)
Comment (nileshpawar1) [reply soap/src/java/com/zimbra/soap/admin/message/ABQDeviceRequest.java]: done

**Pull request merged**